components/profile/heartbeats.rs

```diff
@@ -11,86 +11,68 @@ use std::collections::HashMap;
 use std::env::var_os;
 use std::error::Error;
 use std::fs::File;
-use std::mem;
 use std::path::Path;
+use std::sync::atomic::{ATOMIC_BOOL_INIT, AtomicBool, Ordering};
 
 
 static mut HBS: Option<*mut HashMap<ProfilerCategory, Heartbeat>> = None;
 
+// unfortunately can't encompass the actual hashmap in a Mutex (Heartbeat isn't Send/Sync), so we'll use a spinlock
+static mut HBS_SPINLOCK: AtomicBool = ATOMIC_BOOL_INIT;
+
+fn lock_and_work<F, R>(work: F) -> R where F: FnOnce() -> R {
```

Comment:
Thanks.  I'm trying out this approach which makes sense.  However, when trying to assign the global `HBS` by way of the closure parameter, I get
```
133 |         lock_and_work(|&mut hbs_opt|
    |                             ------- first assignment to `hbs_opt`
134 |             hbs_opt = Some(Box::into_raw(hbs))
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ re-assignment of immutable variable
```
The only way I have around this is to move the `init` and `cleanup` functions to the private module and still assign `HBS` directly using an unsafe block.  So the old approach works for these two functions, but doesn't completely limit access to `HBS` to the `lock_and_work` function, which is what I think we were after.  For reference within `lock_and_work`, I do:
```
work(&mut HBS)
```